src/lib/indexer/index.ts

```diff
@@ -6,31 +6,52 @@ import DefaultBlockHandler from './handleBlock.js'
 import { ChangeSet } from './changeSet.js'
 import { HEX } from '../../models/strings.js'
 import { DbBlock } from '../db/index.js'
+import { injectable, singleton } from 'tsyringe'
+import { serviceState, Status } from '../service-watcher/statusPoll.js'
+import { Env } from '../../env.js'
 
 export type BlockHandler = (blockHash: HEX) => Promise<ChangeSet>
 
 export interface IndexerCtorArgs {
   db: Database
   logger: Logger
   node: ChainNode
+  startupTime: Date
+  env: Env
   handleBlock?: BlockHandler
   retryDelay?: number
 }
 
+export interface BlockProcessingTimes {
+  startupTime: Date
+  lastProcessedBlockTime: Date | null
+  lastUnprocessedBlockTime: Date | null
+}
+
+@singleton()
+@injectable()
 export default class Indexer {
   private logger: Logger
+  private env: Env
   private db: Database
   private node: ChainNode
   private gen: AsyncGenerator<string | null, void, string>
   private handleBlock: BlockHandler
   private retryDelay: number
+  private startupTime: Date
+  private lastProcessedBlockTime: Date | null // while we are running and up to date
+  private lastUnprocessedBlockTime: Date | null // only for when we are catching up on unprocessed blocks
 
-  constructor({ db, logger, node, handleBlock, retryDelay }: IndexerCtorArgs) {
+  constructor({ db, logger, node, handleBlock, retryDelay, startupTime, env }: IndexerCtorArgs) {
     this.logger = logger.child({ module: 'indexer' })
+    this.env = env
     this.db = db
     this.node = node
     this.gen = this.nextBlockProcessor()
     this.retryDelay = retryDelay || 1000
+    this.lastProcessedBlockTime = null
+    this.lastUnprocessedBlockTime = null
+    this.startupTime = startupTime
```

Comment:
can we anotate types on these please. At the moment I assume they're being infered as `any`. The first 2 should be `Date | null` the third just `Date`

Reply:
so they already have these types:
```
  private startupTime: Date
  private lastProcessedBlockTime: Date | null // while we are running and up to date
  private lastUnprocessedBlockTime: Date | null // only for when we are catching up on unprocessed blocks
```

Reply:
yep my mistake